### ISSUE TITLE
Unexpected behaviour with $elemMatch, $size, and exact object match

### ISSUE THREAD

**Issue by jhmajjh** [crcn/sift.js#211]
Hi Craig,

I encountered some unexpected behaviour when querying arrays of embedded objects with **sift 13.3.1** (I'm working on **node 12.16.3**). Say, I have four objects, each with an array of embedded objects. In all my examples, I want to filter depending on what embedded objects there are inside the array.

```
const sift = require('sift')

const values = [
  {
    name: 'zeroElements',
    myArray: []
  },

  {
    name: 'oneElement',
    myArray: [
      { firstKey: 'a', secondKey: 'b' }
    ]
  },

  {
    name: 'twoElements',
    myArray: [
      { firstKey: 'a', secondKey: 'b' },
      { firstKey: 'c', secondKey: 'd' }
    ]
  },

  {
    name: 'otherElement',
    myArray: [
      { firstKey: 'e', secondKey: 'f' }
    ]
  }
]
```

If I run `sift(query)`, the return value is a filter function that returns `true` or `false`, depending on whether the object matches the query, right? But this function appears to change when I use the same function again:

```
const query1 = {
  myArray: {
    $elemMatch: { firstKey: 'a', secondKey: 'b' }
  }
}

console.log('query1 (one sift):', values.filter(sift(query1)))
console.log('query1 (many sifts):', values.filter(x => sift(query1)(x)))
```

Expected behaviour: In both cases `query1` should return `oneElement`, `twoElements`.
Actual behaviour: `query1 (one sift)` returns `oneElement`, `twoElements`, `otherElement`; `query1 (many sifts)` returns `oneElement`, `twoElements`, as expected.

The only difference between the two is that in the first version, `sift(query1)` is called only once, whereas in the second version, `sift(query1)` is called on every element of `values`.

In the following examples, the behaviour is the same, regardless of one sift or many sifts. Here, I want to get the value where `myArray` has only `{ firstKey: 'a', secondKey: 'b' }`, and nothing else.

```
const query2 = {
  myArray: {
    $elemMatch: { firstKey: 'a', secondKey: 'b' },
    $size: 1
  }
}

console.log('query2:', values.filter(sift(query2)))
```

Expected behaviour: `query2` should return `oneElement`.
Actual behaviour: `query2` returns an empty array.

Alternatively, I can make `myArray` exactly match the array `[ { firstKey: 'a', secondKey: 'b' } ]`, but in this case the order of `firstKey` and `secondKey` is important.

```
query3 = {
  myArray: [ { firstKey: 'a', secondKey: 'b' } ]
}

console.log('query3:', values.filter(sift(query3)))
```

Expected behaviour: `query3` returns `oneElement`.
Actual behaviour: `query3` returns `oneElement`. Yay!

Since in MongoDB the field order matters when matching the exact document, and not with `$elemMatch`, the following query should not work:

```
query4 = {
  myArray: [ { secondKey: 'b', firstKey: 'a' } ]
}

console.log('query4:', values.filter(sift(query4)))
```

Expected behaviour: `query4` returns an empty array.
Actual behaviour: `query4` returns `oneElement`.

Best wishes and thank you for your work!

Christopher

**Comment by stalniy** [comment]
Very likely #212 is the same issue

**Comment by crcn** [comment]
Looks like @stalniy  was right - this should be fixed in `13.3.4`.
